test/jdk/java/nio/channels/DatagramChannel/SelectWhenRefused.java

```diff
@@ -50,53 +57,188 @@ public static void main(String[] args) throws IOException {
             dc.register(sel, SelectionKey.OP_READ);
 
             /* Test 1: not connected so ICMP port unreachable should not be received */
-            sendDatagram(dc, refuser);
-            int n = sel.select(2000);
-            if (n > 0) {
-                sel.selectedKeys().clear();
-                // BindException will be thrown if another service is using
-                // our expected refuser port, cannot run just exit.
-                DatagramChannel.open().bind(refuser).close();
-                throw new RuntimeException("Unexpected wakeup");
+            for (int i = 1; i <= MAX_TRIES; i++) {
+                if (testNoPUEBeforeConnection(dc, refuser, sel, i)) {
+                    break; // test succeeded
+                }
+                assertNotEquals(i, MAX_TRIES, "testNoPUEBeforeConnection: too many retries");
             }
 
             /* Test 2: connected so ICMP port unreachable may be received */
             dc.connect(refuser);
             try {
-                sendDatagram(dc, refuser);
-                n = sel.select(2000);
-                if (n > 0) {
-                    sel.selectedKeys().clear();
-                    try {
-                        n = dc.read(ByteBuffer.allocate(100));
-                        throw new RuntimeException("Unexpected datagram received");
-                    } catch (PortUnreachableException pue) {
-                        // expected
+                for (int i = 1; i <= MAX_TRIES; i++) {
+                    if (testPUEOnConnect(dc, refuser, sel, i)) {
+                        break; // test passed
                     }
+                    assertNotEquals(i, MAX_TRIES, "testPUEOnConnect: too many retries");
                 }
             } finally {
                 dc.disconnect();
             }
 
             /* Test 3: not connected so ICMP port unreachable should not be received */
-            sendDatagram(dc, refuser);
-            n = sel.select(2000);
-            if (n > 0) {
-                throw new RuntimeException("Unexpected wakeup after disconnect");
+            for (int i = 1; i <= MAX_TRIES; i++) {
+                if (testNoPUEAfterDisconnect(dc, refuser, sel, i)) {
+                    break; // test passed
+                }
+                assertNotEquals(i, MAX_TRIES, "testNoPUEAfterDisconnect: too many retries");
             }
-
-        } catch(BindException e) {
+        } catch (BindException e) {
```

Comment:
Maybe we should actually log that:

    System.out.println("Skipping test: refuser port has been reused: " + e);